### PR TITLE
Implemented video recording capabilities

### PR DESCRIPTION
**Pull request by joluet**
**Removed Camera1 support**
- Increased `minSdkVersion` to 21
- Deleted `Camera1` class as it uses legacy Camara API that we don't need to support

**Added video recording**
- Added permission for audio recording
- Extended `Camera2` class to allow video recording

Usage:

``` xml
<com.google.android.cameraview.CameraView android:id="@+id/camera"
                                              android:layout_width="match_parent"
                                              android:layout_height="wrap_content"
                                              android:adjustViewBounds="true"
                                              android:keepScreenOn="true"
                                              app:aspectRatio="4:3"
                                              app:autoFocus="true"
                                              app:facing="front"
                                              app:flash="off"
                                              app:minVideoHeight="480"
                                              app:minVideoWidth="640"
                                              app:videoEncodingBitRate="5000000"
                                              app:videoFrameRate="30"/>
```

``` java
cameraView.startRecordingVideo(videoPath);
...
cameraView.stopRecordingVideo();
```

### CODE REVIEW

File: library/src/main/camera2/com/google/android/cameraview/Camera2.java [hunk 1]
```diff
@@ -218,6 +230,20 @@ public void onImageAvailable(ImageReader reader) {
 
     private int mDisplayOrientation;
 
+    private boolean mStartVideoRecording = false;
+
```
Comment (emmaguy) [inline library/src/main/camera2/com/google/android/cameraview/Camera2.java:237]: Lots of whitespace between all these members, seems a bit unnecessary?

File: library/src/main/camera2/com/google/android/cameraview/Camera2.java [hunk 2]
```diff
@@ -478,15 +585,46 @@ private void startCaptureSession() {
...
+            } else {
+                mCamera.createCaptureSession(Arrays.asList(surface, mImageReader.getSurface()),
+                                             mSessionCallback, null);
+            }
         } catch (CameraAccessException e) {
             throw new RuntimeException("Failed to start camera session");
         }
     }
 
+    private android.util.Size chooseVideoSize(android.util.Size[] choices, int width, int height) {
```
Comment (emmaguy) [inline library/src/main/camera2/com/google/android/cameraview/Camera2.java:626]: Static import for this?

File: library/src/main/camera2/com/google/android/cameraview/Camera2.java [hunk 3]
```diff
@@ -478,15 +585,46 @@ private void startCaptureSession() {
...
+        if (bigEnough.size() > 0) {
+            return Collections.min(bigEnough, new Comparator<android.util.Size>() {
+                @Override
+                public int compare(android.util.Size size1, android.util.Size size2) {
+                    return Integer.compare(size1.getWidth() * size1.getHeight(), size2.getWidth() * size2.getHeight());
+                }
+            });
+        } else {
+            Log.e(TAG, "Couldn't find any suitable video recording size");
+            return choices[0];
```
Comment (emmaguy) [inline library/src/main/camera2/com/google/android/cameraview/Camera2.java:646]: is it possible that `choices` is empty?

**Review by emmaguy — APPROVED** [review]
Some very minor comments but otherwise looking good!

**Comment by ivacf** [comment]
Looks good! I'll point my branch to this as soon as the pull request gets merged

File: library/src/main/camera2/com/google/android/cameraview/Camera2.java (hunk 1 above)
Comment (joluet) [reply library/src/main/camera2/com/google/android/cameraview/Camera2.java:237]: Yeah but I wanted to keep the original formatting to minimize the amount of changed lines

File: library/src/main/camera2/com/google/android/cameraview/Camera2.java (hunk 2 above)
Comment (joluet) [reply library/src/main/camera2/com/google/android/cameraview/Camera2.java:626]: They use is a custom Size class, too. To be consistent with the rest I used the package name  to specify I wanna use the one from utils. They've done that, too.

**Comment by emmaguy** [comment]
LGTM 👍

File: library/src/main/camera2/com/google/android/cameraview/Camera2.java (hunk 3 above)
Comment (joluet) [reply library/src/main/camera2/com/google/android/cameraview/Camera2.java:646]: I added this: 

```
if(choices == null || choices.length == 0) {
  throw new RuntimeException("Failed to start video recording session. Camera didn't return any output sizes.");
}
```

I think the camera should return at least one supported output size.

Comment (emmaguy) [reply library/src/main/camera2/com/google/android/cameraview/Camera2.java:646]: Cool, that seems pretty reasonable 👍

**Pull request merged**